src/query/storage/types.go

```diff
@@ -291,15 +297,100 @@ type Querier interface {
 // WriteQuery represents the input timeseries that is written to the database.
 // TODO: rename WriteQuery to WriteRequest or something similar.
 type WriteQuery struct {
+	// opts as a field allows the options to be unexported
+	// and the Validate method on WriteQueryOptions to be reused.
+	opts WriteQueryOptions
+}
+
+// WriteQueryOptions is a set of options to use to construct a write query.
+// These are passed by options so that they can be validated when creating
+// a write query, which helps knowing a constructed write query is valid.
+type WriteQueryOptions struct {
 	Tags       models.Tags
 	Datapoints ts.Datapoints
 	Unit       xtime.Unit
 	Annotation []byte
 	Attributes Attributes
 }
 
+// Validate will validate the write query options.
+func (o WriteQueryOptions) Validate() error {
```

Comment:
nit: this is a `types.go`, maybe pull this out into its own file?

Reply:
Sure thing.